### PR TITLE
fix(fslib): ensure path is a string before calling replace().

### PR DESCRIPTION
**Pull request by nmoinvaz**
**What's the problem this PR addresses?**

Random error when calling `yarn run build` on Windows 11, node 17.6.0, and yarn 3.2.0.
```
Failed to compile.

p.replace is not a function
```
Debugging the issue, here is the value of `p` and also the stack trace during the failure.
```
p = 
<Buffer 5c 5c 3f 5c 43 3a 5c 55 73 65  ... 10 more bytes>
Trace: 
    at Object.toPortablePath (.pnp.cjs:20132:15)
    at PosixFS.mapToBase (.pnp.cjs:22644:18)
    at PosixFS.rmdirSync (.pnp.cjs:22561:39)
    at URLFS.rmdirSync (.pnp.cjs:22561:24)
    at _rmdirSync (node:internal/fs/rimraf:260:21)
    at rimrafSync (node:internal/fs/rimraf:193:7)
    at node:internal/fs/rimraf:253:9
    at Array.forEach (<anonymous>)
    at _rmdirSync (node:internal/fs/rimraf:250:7)
    at rimrafSync (node:internal/fs/rimraf:193:7)
```

**How did you fix it?**

`path` in `fs.rmdirSync` and other Node functions can be `string | Buffer | URL`.  So call `.toString()` before calling `.replace()`.

**Checklist**

- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

- [x] I have set the packages that need to be released for my changes to be effective.

- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

**Review by merceyz — CHANGES_REQUESTED** [review]
Please add a test to https://github.com/yarnpkg/berry/blob/c9ddb263c1c003c6df9cbcaf4e427847edca0d35/packages/yarnpkg-fslib/tests/npath.test.ts

**Comment by nmoinvaz** [comment]
Any idea why the CI fails?

**Comment by nmoinvaz** [comment]
Tests added.

**Review by arcanis — APPROVED** [review]
LGTM

**Review by merceyz — CHANGES_REQUESTED** [review]
This change isn't safe to land (yet?), it causes code that used to throw to now return incorrect data.

Without this PR the following code throws but with this PR it logs an array of strings which is incorrect, it should be an array of `Buffer`s
```js
console.log(require(`fs`).readdirSync(Buffer.from(`.`), { encoding: `buffer` }));
```

Ref https://github.com/yarnpkg/berry/issues/1818

> Any idea why the CI fails?

A flaky test, you can ignore it.

File: packages/yarnpkg-fslib/sources/path.ts [hunk 1]
```diff
@@ -137,11 +137,11 @@ function fromPortablePath(p: Path): NativePath {
 
 // Path should look like "N:/berry/scripts/plugin-pack.js"
 // And transform to "/N:/berry/scripts/plugin-pack.js"
-function toPortablePath(p: Path): PortablePath {
+function toPortablePath(p: Path | Buffer): PortablePath {
   if (process.platform !== `win32`)
     return p as PortablePath;
 
-  p = p.replace(/\\/g, `/`);
+  p = p.toString().replace(/\\/g, `/`);
```
Comment (merceyz) [inline packages/yarnpkg-fslib/sources/path.ts:144]: Forcing it to a string will hide issues like https://github.com/yarnpkg/berry/issues/1889 so this should at least be behind a `Buffer.isBuffer` check.

**Review by merceyz — CHANGES_REQUESTED** [review]

File: packages/yarnpkg-fslib/sources/path.ts (hunk 1 above)
Comment (nmoinvaz) [reply packages/yarnpkg-fslib/sources/path.ts]: Good idea. Fixed.

**Comment by nmoinvaz** [comment]
Anything else on this commit to do?

**Comment by paul-soporan** [comment]
I apologize for the delay, I'll try doing an in-depth review tomorrow.

**Review by paul-soporan — CHANGES_REQUESTED** [review]
Haven't had time to look into the encoding issues yet, but, regarding the API, `toPortablePath` isn't the right place for this change. It's only meant to accept a `Path`.

Introducing an invisible layer here means that things might break (Edit: Actually, in this specific case they wouldn't, but still, it gives off the false impression that we fully support `Buffer`s without them being implemented in the right layer) when going through the `FS` classes that only expect paths.

The right fix would be to create a `BufferFS` to automatically wrap the `patchedFs` just like the [`URLFS`](https://github.com/yarnpkg/berry/blob/master/packages/yarnpkg-fslib/sources/URLFS.ts) does.

File: packages/yarnpkg-fslib/sources/path.ts [hunk 2]
```diff
@@ -138,10 +138,13 @@ function fromPortablePath(p: Path): NativePath {
 
 // Path should look like "N:/berry/scripts/plugin-pack.js"
 // And transform to "/N:/berry/scripts/plugin-pack.js"
-function toPortablePath(p: Path): PortablePath {
+function toPortablePath(p: Path | Buffer): PortablePath {
   if (process.platform !== `win32`)
     return p as PortablePath;
 
+  if (Buffer.isBuffer(p))
+    p = p.toString();
```
Comment (paul-soporan) [inline packages/yarnpkg-fslib/sources/path.ts:146]: (Just for reference, this change is redundant with the other change I requested) Having the buffer check after the platform check means that `Buffer`s would still not be supported on unix.

Comment (nmoinvaz) [reply packages/yarnpkg-fslib/sources/path.ts]: Ok

**Comment by nmoinvaz** [comment]
I have made the suggested changes.

**Comment by paul-soporan** [comment]
I'm sorry if that wasn't clear - I meant to say that we should create a `BufferFS` class inside a new `BufferFS` file (just like [`URLFS`](https://github.com/yarnpkg/berry/blob/master/packages/yarnpkg-fslib/sources/URLFS.ts) is implemented), not modify `PosixFS`. The `patchFs` function should then be modified to wrap the existing fs with a `BufferFS` too (https://github.com/yarnpkg/berry/blob/e143651fc178cd830b91b9bd2e9c93e11c3bacdf/packages/yarnpkg-fslib/sources/patchFs.ts#L146-L147).

**Comment by nmoinvaz** [comment]
I have closed the pull request. I don't want to see it through.

My workaround is just to delete the build directory before running yarn build.

**Pull request closed without merging**